File: bindings/python/examples/pipeline/encode_to_mp4.py
"""Auto-pick the best available H.264 encoder, push 90 synthetic grey NV12
frames through it, and write the resulting fragmented MP4 to out.mp4.

✅ REAL — the auto video encode -> fMP4 pipeline is implemented in the native
C ABI (`mediaway_pipeline_ffi`); this example runs against it. Only the frames
themselves are synthetic (flat grey, NV12).

The encoder is chosen for us by AutoVideoEncoder.pick(). EncodeSession takes
ownership of that choice, and finish() is terminal: it returns the complete
MP4 bytes and consumes the session (no close() afterwards — __exit__ on an
already-finished session is a no-op).
"""

from pathlib import Path

from mediaway import (
    AutoVideoEncoder,
    Codec,
    EncodeSession,
    EncoderUnavailableError,
    PixelFormat,
    Rational,
    VideoFrame,
)

WIDTH = 640
HEIGHT = 480
FRAMES = 90


def make_grey_nv12_frame(i: int) -> VideoFrame:
    """A flat-grey NV12 frame (Y = U = V = 0x80) with presentation time i/30 s."""
    y_plane = WIDTH * HEIGHT
    uv_plane = WIDTH * HEIGHT // 2
    return VideoFrame(
        width=WIDTH,
        height=HEIGHT,
        format=PixelFormat.NV12,
        data=bytes([0x80]) * (y_plane + uv_plane),
        pts=Rational(i, 30),
    )


def main() -> None:
    try:
        encoder = AutoVideoEncoder.pick(
            codec=Codec.H264,
            width=WIDTH,
            height=HEIGHT,
            frame_rate=Rational(1, 30),
        )
    except EncoderUnavailableError as err:
        print(f"no H.264 encoder available: {err}")
        return
    print(f"picked encoder: {encoder.name} ({encoder.codec.name})")

    with EncodeSession(encoder) as session:
        for i in range(FRAMES):
            session.push_frame(make_grey_nv12_frame(i))
        mp4 = session.finish()  # terminal: consumes the session

    out_path = Path("out.mp4")
    out_path.write_bytes(mp4)
    print(f"encoded {FRAMES} frames -> {out_path} ({len(mp4)} bytes)")


if __name__ == "__main__":
    main()

File: bindings/nodejs/examples/pipeline/encode-to-mp4.ts
/**
 * encode-to-mp4.ts — auto video encode → fragmented MP4
 * Mirrors: examples/pipeline/encode_to_mp4.rs
 *
 * Status: ✅ real ABI under it.
 * The C ABI's pipeline capability (auto video encode → fMP4, mediaway_pipeline_ffi)
 * is implemented: one call picks the best available OS/GPU encoder for the config and
 * wires it into an internal MP4 muxer. Video only — no audio encoder in the ABI.
 *
 * Flow: encode 90 synthetic grey NV12 frames with the auto-selected H.264 encoder and
 * write the resulting MP4 bytes to out.mp4. When no encoder backend exists on this
 * machine we catch EncoderUnavailableError and exit cleanly.
 *
 * Run (once @mediaway/* packages exist): npx tsx examples/encode-to-mp4.ts
 */

import { writeFileSync } from "node:fs";

import {
  AutoVideoEncodeConfig,
  EncodeSession,
  VideoFrame,
  openAutoEncoder,
  EncoderUnavailableError,
  MediawayError,
} from "@mediaway/encoder";
import { Rational } from "@mediaway/container";

const CODEC = "h264";
const WIDTH = 640;
const HEIGHT = 480;
const TIME_BASE: Rational = { num: 1, den: 30 };
const FRAME_COUNT = 90; // 3 s at 30 fps
const OUT = "out.mp4";

async function main(): Promise<void> {
  const config = AutoVideoEncodeConfig.defaults(CODEC, WIDTH, HEIGHT, TIME_BASE);
  config.bitrateBps = 1_500_000; // override the backend default for this resolution

  let session: EncodeSession;
  try {
    session = await openAutoEncoder(config);
  } catch (err) {
    if (err instanceof EncoderUnavailableError) {
      console.log(`no ${CODEC} encoder backend on this machine; nothing to do`);
      return;
    }
    throw err;
  }

  try {
    // Grey NV12: Y plane and both chroma planes at 128 → neutral grey, no gradients.
    const greyFrame = Buffer.alloc((WIDTH * HEIGHT * 3) / 2, 0x80);

    for (let i = 0; i < FRAME_COUNT; i++) {
      const frame: VideoFrame = {
        pts: i, // units of the config's timeBase (1/30 s)
        duration: 1,
        width: WIDTH,
        height: HEIGHT,
        pixelFormat: "nv12",
        data: greyFrame, // safe to reuse: the wrapper copies per writeFrame call
      };
      await session.writeFrame(frame);
    }

    // finish() is terminal: it consumes the session and returns the complete MP4.
    const mp4 = await session.finish();
    writeFileSync(OUT, mp4);
    console.log(`encoded ${FRAME_COUNT} frames (${CODEC} ${WIDTH}x${HEIGHT}) -> ${OUT} (${mp4.length} bytes)`);
  } finally {
    // Idempotent; a no-op after finish() consumed the session, but required to free
    // the native handle when an error aborted the encode mid-way.
    session.close();
  }
}

main().catch((err) => {
  if (err instanceof MediawayError) {
    console.error(`mediaway error (status ${err.status}): ${err.message}`);
  } else {
    console.error(err);
  }
  process.exitCode = 1;
});
